test/jit/test_save_load.py

```diff
@@ -983,3 +983,46 @@ def forward(self) -> Optional[FooTuple]:
         m_loaded = self.getExportImportCopy(torch.jit.script(MyModule()))
         output = m_loaded()
         self.assertEqual(output, None)
+
+    def test_save_load_params_buffers_submodules(self):
+        """
+        Check that parameters, buffers, and submodules are the same after loading.
+        """
+
+        class Submodule(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+
+        class TestModule(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.add_module("submodule_a", Submodule())
+                self.register_parameter("parameter_a", torch.nn.Parameter(torch.randn(4)))
+                self.register_buffer("buffer", torch.randn(4))
+                self.t = torch.rand(4) # not buffer
+
+                self.parameter_b = torch.nn.Parameter(torch.randn(4))
+                self.submodule_b = Submodule()
+
+        m = TestModule()
+        m_loaded = self.getExportImportCopy(torch.jit.script(m))
+
+        # Check submodules.
+        self.assertEqual(len(list(m.named_modules())), len(list(m_loaded.named_modules())))
+        for m_s, loaded_s in zip(m.named_modules(), m_loaded.named_modules()):
+            m_name, _ = m_s
+            loaded_name, _ = loaded_s
+            self.assertEqual(m_name, loaded_name)
+
+        # Check parameters.
+        self.assertEqual(len(list(m.parameters())), len(list(m_loaded.parameters())))
+        for m_p, loaded_p in zip(m.parameters(), m_loaded.parameters()):
+            self.assertEqual(m_p, loaded_p)
+
+        # Check buffers.
```

Comment:
For this test case, we should see both attributes and buffer as "buffer" right?

Reply:
Yes, that's what TorchScript returns but my claim is that this is not correct. Eager mode output for this case is:
```
>>> class Mod(torch.nn.Module):
...   def __init__(self):
...     super().__init__()
...     self.t = torch.rand(4)
...     self.register_buffer("buffer", torch.rand(4))
... 
>>> m = Mod()
>>> list(m.buffers())
[tensor([0.0439, 0.5317, 0.5104, 0.7503])]
>>> list(m.named_buffers())
[('buffer', tensor([0.0439, 0.5317, 0.5104, 0.7503]))]
```
The diff stacked on top fixes this problem by applying the change to `BufferPolicy` that you reverted.

Reply:
Sounds great, which diff should land first? I was thinking we should land this first (even if the behavior is not right), but if we want to land the other one that's fine with me too. Also please import the other diff so that I can try it out :)

Reply:
I think both need to land at once - this diff will break CI if landed by itself.

#49588 has already been imported as D25633250.